src/components/mobile-menu/mobile-menu.jsx

```diff
@@ -0,0 +1,153 @@
+
+  render() {
+    return (
+      <React.Fragment>
+        <SwipeEventEmitter
+          onRelease={this.onSwipeRelease}
+          onHorizontalMove={this.updateDrawerPosition}
+        />
+        <Flex justifyContent="space-between" alignItems="center">
+          {this.props.renderLogo()}
+          <MenuIcon onToggle={this.toggle} isOpened={this.state.isMenuOpen} />
+        </Flex>
+        <Portal>
+          <Spring
+            to={this.animationPose}
+            immediate={name =>
+              this.state.drawerPosition !== 0 && name === 'translation'
+            }>
+            {({ translation, opacity }) => (
+              <MobileSidebar
+                translation={translation}
+                transitioned={this.state.drawerPosition === 0}>
+                {this.isSticky ? (
+                  <Flex
+                    justifyContent="space-between"
+                    alignItems="center"
+                    m={'2rem'}
+                    style={{
+                      opacity,
+                      position: 'absolute',
+                      top: 0,
+                      left: 0,
+                      right: 0,
+                    }}>
+                    {this.props.renderLogo()}
+                    <MenuIcon
+                      onToggle={this.toggle}
+                      isOpened={this.state.isMenuOpen}
+                    />
+                  </Flex>
+                ) : null}
+                {this.props.children}
+              </MobileSidebar>
+            )}
+          </Spring>
+        </Portal>
+      </React.Fragment>
+    )
+  }
+  get animationPose() {
+    const { isMenuOpen, drawerPosition } = this.state
+    let windowWidth = window.innerWidth
+    let finalDistance = ''
+    if (!isMenuOpen) {
+      finalDistance = drawerPosition
+    }
+    if (isMenuOpen) {
+      finalDistance = -windowWidth
+    }
+    if (isMenuOpen && drawerPosition > 0) {
+      finalDistance = -windowWidth + drawerPosition
+    }
+    return {
+      translation: finalDistance,
+      opacity: -finalDistance / windowWidth,
+    }
+  }
+  get isSticky() {
+    return this.props.stickyOffset < window.pageYOffset
+  }
+  updateDrawerPosition = event => {
+    this.setState(pState => ({ ...pState, drawerPosition: event.distance }))
+  }
+  onSwipeRelease = event => {
+    const { distance } = event
+    this.setState(pState => ({
+      ...pState,
+      drawerPosition: 0,
+    }))
+    if (distance < -130) {
+      return this.showMenu()
+    }
+    if (distance > 130) {
+      return this.hideMenu()
+    }
+  }
+  showMenu = () => {
+    document.body.style.overflow = 'hidden'
+    return this.setState(pState => ({
+      ...pState,
+      isMenuOpen: true,
+    }))
+  }
+  hideMenu = () => {
+    document.body.style.overflow = 'visible'
+    return this.setState(pState => ({
+      ...pState,
+      isMenuOpen: false,
+    }))
+  }
+  toggle = () => {
+    this.state.isMenuOpen ? this.hideMenu() : this.showMenu()
```

Comment:
and here is incorrect usage of `setState`

you probably need something like this:

```
toggle = () => this.setState(({ isMenuOpen }) => { isMenuOpen: !isMenuOpen  })
```

or even:

```
toggle = val => this.setState(({ isMenuOpen }) => { isMenuOpen: val != null ? val : !isMenuOpen  })
```

and you wont need hide/show functions

Reply:
Oh no, actually first of all I *do* need both show and hide methods, because on swipe hideMenu could be called several times in a row. And also each of these two function do not only call setState, but also do some imperative side effects. And if you ask what is better to write this.toggle(true) or this.showMenu(), I will go for the second option

Reply:
you still can create `hide`/`show` methods using `toggle` method:

```
show: () => this.toggle(true)
```

```
hide: () => this.toggle(false)
```

Reply:
And where should I do side effects `document.body.style.overflow = 'hidden'`? inside toggle? in this case we need to check for variable presence, then for current value, and based on these do this or that stuff. Sorry, but I still think that it is unnecessary complexity.

Reply:
I've just missed `document.body.style.overflow = 'hidden'`

you can place it in the `hide`/`show` functions

the whole point was in incorrect usage of set state

Reply:
and in these case I need to once again set document styles inside toggle function... 

During development I do as you suggest, but as I've said I think it's unnecessary complexity.

Reply:
Why do you say that these is incorrect? Does it cause bugs? or some performance issues? I think no. But repeating code is bad. Do complex branching logic is bad. At least for me.